### PR TITLE
Feature/7 new wallet named wallets api

### LINKED ISSUE #7: New wallet, named wallets (+basic REST API)
Though, of course, anyone is free to take this task, but I'd prefer @2qx to take it, since he seems to have a lot of understanding how to do this well and this task is the basis of ALL of our future work. Don't worry, as soon as we're done with some basic stuff, there will be plenty of tasks.

This issue calls to implement the creation of a new wallet and its corresponding REST API:

```
const w = new RegTestWallet();
console.log(w.wif);
// cNfsP... (random every time)
```

and named wallets:

```
const w1 = new RegTestWallet('username1');
console.log(w1.wif);
// abcdef.. (random first time)

const w2 = new RegTestWallet('username1');
console.log(w2.wif);
// abcdef.. (same as above)

const w2 = new RegTestWallet('username2');
console.log(w2.wif);
// fghi.. (random)
```

It should store its private key (in WIF format I guess?) in IndexedDB in the browser by name. The usefulness of this comes when the same browser is used by different people (or one person) logging in to different accounts on the same site. i.e. on read.cash you can have multiple usernames, each one having its own wallet. Without this any person who tries to create a new account on any site will end up with the same wallet on every account (which is incorrect). read.cash doesn't actually store it by username, but rather by a hash of user.id + secret string, so that it's harder to guess. But it will be just a recommendation.

Also, we need to start doing the node.js interface, so similarly we need to implement 

```
POST /api/v1/wallet
JSON request body: {"type":"wif", "network":"regtest"}
```

returning JSON:
```
{"wallet": "wif:regtest:abc12345", "wif": "abc12345", "cashaddr": "..."}
```

This call is stateless, so it doesn't save anything, just returns a random wallet. `abc12345` is the actual random WIF, like `cNfsPtqN2bMRS7vH5qd8tR8GMvgXyL5BjnGAKgZ8DYEiCrCCQcP6`.

....and a named version of the call:

```
POST /api/v1/wallet
{"type":"wif", "network":"regtest", "name": "username1"}
```

returning
```
{"wallet": "wif:regtest:username1", "wif": "abc12345", "cashaddr": "..."}
```

But this call is stateful, meaning each time user calls for `{"type":"wif", "network":"regtest", "name": "username1"}` he gets the same wallet back.

`wallet` in return is the [serialization](https://github.com/mainnet-cash/project/issues/2) of the wallet data that we would need in future calls - it's important for us to know that this is a WIF wallet (as opposed to HD wallets), that it's for regtest and what its WIF is... for named wallets we don't need to store WIF in the `wallet`, since we store WIF on server.

So, now a future call of 

```
POST /api/v1/send
{"wallet": "wif:regtest:abcd123", "to" ...}
```

...we already know what to do.

Some considerations for storage: for browser it's surely IndexedDB (though https://dexie.org/ looks like a much nicer wrapper API and pretty small). 

For server-side node.js, I have no idea what we should use. I would probably avoid using sqlite, because some servers could see heavy use, that means having to run multiple workers of node.js, and SQLite won't work well with that. So maybe it's PostgreSQL or mySQL.. This looks interesting though: https://github.com/indexeddbshim/IndexedDBShim

Yeah and I guess if we're talking Postgres/mySQL - then we'll need docker-compose anyway.

Tests are required.

### -- $1,000 bounty --

The bounty is paid in BCH from the funds collected via https://github.com/mainnet-cash/mainnet-js . To claim the bounty, please do the following: 

1) if this task requires coding in TypeScript, first you need to complete the test task by sending a private message to ReadCash via [Slack link here](https://github.com/mainnet-cash/project#where-are-we); 

2) after you have completed the test task once, tell us that you are working on this task and how long it is going to take, like this: "I'll take it. I expect it to be done in 2 days". If that time close to expire and you haven't finished it yet, you can write something like "It'll take 2 more days", otherwise anyone else is free to take the task); 

3) send a Pull Request or publish the solution here as described in the issue. 
Good luck! If you think that bounty is way too small, please notify me about it and we can discuss. A bounty will be paid via the chaintip bot.

### PR DESCRIPTION
**Pull request by 2qx**
Breaking this down into sub-tasks:

- [x] Codify API contract
- [x] Generate services for Express Server
- [x] Implement generateWif() in core, and express-server 
- [x]  Preliminary webpack versions for node, browser and webworkers
- [x]  Implement schema for indexeddb in browsers, with a simple test

**Note** The express server has been refactored to typescript.  

closes #7

### CODE REVIEW

File: serve/combined.log [hunk 1]
```diff
@@ -0,0 +1,50 @@
+{"message":"Express server running","level":"info","service":"user-service","timestamp":"2020-08-21T15:33:34.348Z"}
```
Comment (readcash) [inline serve/combined.log:1]: We probably need to remove and .gitignore this file

Comment (2qx) [reply serve/combined.log]: Added `serve/*.log` to .gitignore and removed both of those files.

File: client/README.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
+This project uses codegen to generate thin-client libraries for the api. These libraries
+handle the boilerplate of connecting to the server for you.
+
+## Clients
+
+| language         | Examples | Tests | Native Signing |
+| :--------------- | :------: | :---: | :------------: |
+| mainnet-node     |    ⍰     |   ⍰   |       👍       |
```
Comment (readcash) [inline client/README.md:8]: I would suggest to rename this to `typescript-node`, `javascript-browser` and I'm not sure what `mainnet-worker` is...

File: jest.node.setup.js [hunk 3]
```diff
@@ -21,4 +25,12 @@ module.exports = async function () {
   } else {
     console.log("...already running");
   }
+  if (global.mainnetServer === undefined) {
+    global.mainnetServer = spawn("npx", ["ts-node", "./serve/index.ts"], {
+      shell: false,
+    });
+  }
+  console.log("Waiting for readiness");
+  await delay(3000);
```
Comment (readcash) [inline jest.node.setup.js:34]: This is usually a bad idea that is bound to break if users computer is slow or a network is slow.. It's better to replace it with some readiness check.

File: serve/api/openapi.yaml [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+          content: {}
+          description: Invalid input
+      summary: create a new wallet
+      tags:
+        - wallet
+      x-codegen-request-body-name: body
+      x-eov-operation-handler: controllers/WalletController
+  /send:
+    get:
+      operationId: getSendNotAllowed
```
Comment (readcash) [inline serve/api/openapi.yaml:68]: So, we remove this, right?

File: client/README.md (hunk 2 above)
Comment (2qx) [reply client/README.md:8]: On the table, I was trying to show which features would be supported by the module builds, i.e. the js module built for node, the js module built for the browser, and the js module built for web workers (browser threads). But I should probably just remove them from that table and instructions in the main README.md

File: serve/api/openapi.yaml (hunk 4 above)
Comment (2qx) [reply serve/api/openapi.yaml:68]: I removed this from the api, and the typescript mock client. 

I can remove this from the express server when I fix the templates.

File: jest.node.setup.js (hunk 3 above)
Comment (2qx) [reply jest.node.setup.js]: fixed

File: serve/api/openapi.yaml (hunk 4 above)
Comment (2qx) [reply serve/api/openapi.yaml:68]: Which I created an issue for here: https://github.com/mainnet-cash/mainnet-js/issues/12

**Pull request merged**